### DISCUSSION TITLE
How to solve TypeScript error: getState does not exist on EnhancedStore?

### DISCUSSION THREAD

**Discussion by silviubogan** [reduxjs/redux-toolkit#1204] [Q&A]
The error:

```log
error TS2339: Property 'getState' does not exist on type 'EnhancedStore<any, any, MiddlewareArray<any>>'.
```

The store is defined like this:

```js
const store = configureStore({
  reducer: rootReducer,
  middleware,
  devTools: process.env.NODE_ENV !== "production",
  enhancers: [reduxBatch],
});
```

Please help me solve this. I just want to obtain the actual store data from the `store` object, not inside a component but inside a setup function.

Thanks!

**Comment by phryneas** [comment]
In this case, either `rootReducer` or `middleware` could be typed incorrectly (they very likely are, judging from the `any`s in your error message). Also we have no idea how the code actually leading to the error looks like. You'll have to provide more context that that.

**Reply by silviubogan** [reply]
```tsx
import { rootReducer /* , rootSaga */ } from "./rootReducer";

[...]

const middleware = getDefaultMiddleware({
    immutableCheck: false,
    serializableCheck: false,
    thunk: true,
  }).concat(idleMiddleware) /* .concat(sagaMiddleware) */;
```

and

```tsx
export const rootReducer = combineReducers({
    [...]
});
```

The error is received when compiling the client (the app). I have an Axios interceptor for dealing with something related to authentication. In it I must access the store data and set a HTTP header based on that.

**Reply by phryneas** [reply]
Move the middleware inline and use the callback notation.
```ts
const store = configureStore({
  reducer: rootReducer,
  middleware: getDefaultMiddleware => getDefaultMiddleware({
    immutableCheck: false,
    serializableCheck: false,
    thunk: true,
  }).concat(idleMiddleware),
  devTools: process.env.NODE_ENV !== "production",
  enhancers: [reduxBatch],
});
```

Also, hover over your rootReducer to make sure it is correctly typed to your state shape and not to `any`.

**Reply by silviubogan** [reply]
`rootReducer` is of type `any` because `combineReducers` from "redux" is not available in "redux" because the "redux" import looks to the "redux" directory in my application, not in node_modules. I am trying to solve this. Thank you!

**Reply by phryneas** [reply]
import it from `@reduxjs/toolkit`. But yeah, generally having a `redux `directly regularlt causes problems, we advise against that.

**Reply by silviubogan** [reply]
Thank you for the answer. I get this error:

> error TS2305: Module '"@reduxjs/toolkit"' has no exported member 'combineReducers'.

I posted a question on TS Discord about how can I make baseUrl in tsconfig.json to only apply to path aliases. Razzle requires me to use "src" as the baseUrl:

> ERROR  Your project's `baseUrl` can only be set to `src` or `node_modules`. Razzle does not support other values at this time

**Reply by phryneas** [reply]
Then even that resolves to that `redux` folder. Best thing you can do is probably renaming that to `store`.

**Reply by silviubogan** [reply]
I renamed src/redux to src/store. Now I have this code:

```ts
const store = configureStore({
  reducer: rootReducer,
  middleware: getDefaultMiddleware({
    immutableCheck: false,
    serializableCheck: false,
    thunk: true,
  }).concat(idleMiddleware),
  devTools: process.env.NODE_ENV !== "production",
  enhancers: [reduxBatch],
});
```

and there is a red squiggle under reduxBatch.

```log
Type 'StoreEnhancer<{ dispatch: (actions: AnyAction[]) => AnyAction[]; }, {}>' is not assignable to type 'StoreEnhancer<{}, {}>'.
  Types of parameters 'next' and 'next' are incompatible.
    Types of parameters 'preloadedState' and 'preloadedState' are incompatible.
      Type 'import("/path/to/my/project/node_modules/redux/index").PreloadedState<S> | undefined' is not assignable to type 'import("/path/to/my/project/node_modules/@reduxjs/toolkit/node_modules/redux/index").PreloadedState<S> | undefined'.
        Type 'PreloadedState<S>' is not assignable to type 'PreloadedState<S> | undefined'.
          Type '(S extends CombinedState<infer S1> ? { [K in keyof S1]?: (S1[K] extends object ? PreloadedState<S1[K]> : S1[K]) | undefined; } : never) | { [K in keyof S]: S[K] extends object ? PreloadedState<...> : S[K]; }' is not assignable to type 'PreloadedState<S> | undefined'.
            Type '{ [K in keyof S]: S[K] extends object ? PreloadedState<S[K]> : S[K]; }' is not assignable to type 'PreloadedState<S>'.
              Type 'import("/path/to/my/project/node_modules/redux/index").PreloadedState<S>' is not assignable to type 'import("/path/to/my/project/node_modules/@reduxjs/toolkit/node_modules/redux/index").PreloadedState<S>'.
                Type '(S extends CombinedState<infer S1> ? { [K in keyof S1]?: (S1[K] extends object ? PreloadedState<S1[K]> : S1[K]) | undefined; } : never) | { [K in keyof S]: S[K] extends object ? PreloadedState<...> : S[K]; }' is not assignable to type 'PreloadedState<S>'.
                  Type '{ [K in keyof S]: S[K] extends object ? PreloadedState<S[K]> : S[K]; }' is not assignable to type 'PreloadedState<S>'.ts(2322)
```

I don't know yet how to read this type of errors, or more specific, I do not trust my ability to deduce solutions from such messages.

I hope it is not about the redux-batch package because it is not recently updated.

**Reply by phryneas** [reply]
Just do a cast there - something like `reduxBatch as unknown as StoreEnhancer<{}, {}>`  should do the trick.

**Reply by silviubogan** [reply]
`dispatch(someActionHere(...))` does not compile because there is red squiggle under `someActionHere`:

```log
Argument of type 'AsyncThunkAction<unknown, ISomeInterface, {}>' is not assignable to parameter of type 'AnyAction'.
  Property 'type' is missing in type 'AsyncThunkAction<unknown, ISomeInterface, {}>' but required in type 'AnyAction'.
```

The tooltip of someActionHere(...) is

```log
const dispatch: Dispatch
<AnyAction>(action: AnyAction) => AnyAction
```

Background:
1. `dispatch` is obtained like this:  `const dispatch = useAppDispatch()`;
2. `useAppDispatch` is defined like this: `export const useAppDispatch = () => useDispatch<AppDispatch>()`
3. `AppDispatch` is defined like this: `export type AppDispatch = typeof store.dispatch;`
4. `store` is defined like this:
```tsx
const store = configureStore({
  reducer: rootReducer,
  middleware: getDefaultMiddleware({
    immutableCheck: false,
    serializableCheck: false,
    thunk: true,
  }).concat(idleMiddleware),
  devTools: process.env.NODE_ENV !== "production",
  enhancers: [reduxBatch as unknown as StoreEnhancer],
});
```

I could cast postLicitation(...) to `any` but I should do it in a lot of places and it is something like a hack.

**Reply by phryneas** [reply]
You are not using the callback notation for middleware.

**Reply by silviubogan** [reply]
Thank you! I changed that. The same error occurs.

**Reply by silviubogan** [reply]
Proof:

```tsx
const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      immutableCheck: false,
      serializableCheck: false,
      thunk: true,
    }).concat(idleMiddleware),
  devTools: process.env.NODE_ENV !== "production",
  enhancers: [(reduxBatch as unknown) as StoreEnhancer],
});
```

**Reply by phryneas** [reply]
Does it change if you skip the idleMiddleware? If it does, that one might by typed wrong

**Reply by silviubogan** [reply]
Thank you for the attempt. Unfortunately, it does not change.

**Reply by phryneas** [reply]
Then please check if your node_modules somewhere contain redux 4.0.5 and redux 4.1.0 at the same time. That seems to cause problems like this.

**Reply by silviubogan** [reply]
![image](https://user-images.githubusercontent.com/198924/122935801-4eb63680-d379-11eb-88be-df862509f485.png)

I hope I can solve this issue by manually upgrading some of the deps of the app. This is what I think I will try.

**Reply by silviubogan** [reply]
Now I realized that I did not print the elements from the entire tree of dependencies in the screenshot above. :sweat_smile:

**Reply by silviubogan** [reply]
![image](https://user-images.githubusercontent.com/198924/122938186-39420c00-d37b-11eb-9c39-ecf4336e22a1.png)

I don't know how to search in the npm list output above what packages should I upgrade. The command I used to output this list is not ideal.

**Reply by silviubogan** [reply]
Hopefully I've learned something :sweat_smile: 

![image](https://user-images.githubusercontent.com/198924/122939200-1bc17200-d37c-11eb-9945-eb6c249b56c3.png)

**Reply by phryneas** [reply]
I only have used `resolutions` in `yarn` before, but maybe this could help you: https://www.npmjs.com/package/npm-force-resolutions

**Reply by silviubogan** [reply]
I still get errors although:

![image](https://user-images.githubusercontent.com/198924/122957556-b759df00-d38a-11eb-9bed-1cf1a3bcb3f1.png)

The errors:

![image](https://user-images.githubusercontent.com/198924/122958202-51ba2280-d38b-11eb-9f98-3ec163dc38a1.png)

I would insist upon this message: https://github.com/reduxjs/redux-toolkit/discussions/1204#discussioncomment-904296

Thank you for all the help!

**Reply by phryneas** [reply]
Yeah.. that `AnyAction` can either come from a middleware (do you still have that problem if you comment them all out?) or from that redux issue :/
What do you get when you hover your AppDispatch type?

**Reply by silviubogan** [reply]
If I remove all middlewares I get only one error:
![image](https://user-images.githubusercontent.com/198924/123077792-0bb09d80-d423-11eb-94cf-3b36cc3a4b11.png)

The AppDispatch has the tooltip of this form when using no middlewares:

```ts
type AppDispatch = ThunkDispatch<CombinedState<{
    [...]: {
        [...]
    } & PersistPartial;
    [...]: {
        [...]: {
            [index: string]: [...];
        };
        loading: boolean;
        error: string | null;
    };
    [...]: {
        ...;
    };
    [...]: {
        ...;
    };
    idle: unknown;
}>, null, AnyAction> & ThunkDispatch<...> & Dispatch<...>
```

When using all the middlewares as I did previously, the tooltip is:

```ts
type AppDispatch = Dispatch<AnyAction>
```

**Reply by silviubogan** [reply]
I tried to configure the store like this:

    middleware: (getDefaultMiddleware) => [idleMiddleware]

and I see the same 15 errors appearing (I solved the other error that appeared alone).

I will have to contribute to this project: https://github.com/noderaider/redux-idle-monitor. I wish I get an answer to [the issue I opened there](https://github.com/noderaider/redux-idle-monitor/issues/31) to start working on it.

Thanks for all the help.

**Reply by phryneas** [reply]
The "complex" `AppDispatch` type there is the correct one. You 100% **need** to use the middleware callback with getDefaultMiddleware().concat. If it does not work with `idleMiddleware`, just do a `idleMiddleware as Middleware`, that should work.

That one remaining error seems to stem from an assignment to a pre-typed variable. In what code is that coming up?

**Reply by silviubogan** [reply]
I was astonished to see that `idleMiddleware as Middleware` makes all the 15 errors go away.

I solved that single error by replacing

```ts
export default function setupAxios(
  axios: AxiosStatic,
  store: EnhancedStore<any, any, MiddlewareArray<any>>
) {
```

with

```ts
export default function setupAxios(
  axios: AxiosStatic,
  store: EnhancedStore
) {
```

So now there are no more errors from RTK. I just have to solve a few with the Razzle server-side code and with the razzle.config.js file in my project.

I've got such a nice kind of support here!

Thank you a lot!
